Boris-Em/office-ui-fabric-react
# Implement Document Title Bar

## Description

The Document Title Bar (DTB) is a component that allows users to rename the current document as well as performing some other file-related commands such as Go To Saved Location, and See Versions.

The DTB works like a Contextual Menu Button because it is a button that can present a contextual menu when clicked. We do not build the DTB on top of the existing Fabric `Button` class because the DTB behaves and looks quite differently from the generic Fabric `Button` to be considered as a "customized button". The menu items in the contextual menu are built-in to the DTB rather than being specified by the consumer because those items define what a DTB is.

In its normal state, the DTB is a button that contains the document title, the document's file status (e.g. Saving, Saved, etc), and a chevron icon that indicates to the users that a contextual menu can become visible when they click on the button.

When the user clicks Rename, it will trigger a React rendering process that changes the button element to an input element. When the input element receives an onBlur event or an Enter-keyDown event, the DTB will invoke the onRenameDocument callback. An Escape-keyDown event will cancel the renaming procedure and not invoke the onRenameDocument callback.

Clicking on Saved Location and Versions will simply invoke their respective callbacks.

#### Notes
- The original changes included necessary modifications to how the `ContextualMenu` class styles its child elements so that we could achieve the new Ribbon's appearance, but we decided to defer those modification till later when there is a more comprehensive plan for custom theming Fabric components. As a result, the DTB's contextual menu will have the default contextual menu's appearance.
- We now will not render an empty space for any Contextual Menu Header item that does not have an icon.
- A consumer of the DTB can decide whether or not to show Versions in the contextual menu by passing a boolean to the hasVersions property.

## Testing

Unit tests for the DTB have been implemented and cover the following scenarios:
- Verify that the DTB displays the specified document title
- Verify that the DTB displays both the document title and the status text
- Verify that the DTB populates contextual menu items correctly
- Verify that the DTB prefills the input field with the current document title when the user clicks Rename
- Verify that callback functions are invoked correctly
- Verify that the DTB does not invoke onRenameDocument when it receives an Escape-keyDown event during rename mode
